hack/generator/pkg/jsonast/schema_abstraction_openapi.go

```diff
@@ -0,0 +1,288 @@
+	if len(schema.inner.Title) == 0 {
+		return nil // translate to optional
+	}
+
+	return &schema.inner.Title
+}
+
+func (schema *OpenAPISchema) url() *url.URL {
+	url, err := url.Parse(schema.inner.ID)
+	if err != nil {
+		return nil
+	}
+
+	return url
+}
+
+func (schema *OpenAPISchema) hasType(schemaType SchemaType) bool {
+	return schema.inner.Type.Contains(string(schemaType))
+}
+
+func (schema *OpenAPISchema) hasAllOf() bool {
+	return len(schema.inner.AllOf) > 0
+}
+
+func (schema *OpenAPISchema) allOf() []Schema {
+	return schema.transformOpenAPISlice(schema.inner.AllOf)
+}
+
+func (schema *OpenAPISchema) hasAnyOf() bool {
+	return len(schema.inner.AnyOf) > 0
+}
+
+func (schema *OpenAPISchema) anyOf() []Schema {
+	return schema.transformOpenAPISlice(schema.inner.AnyOf)
+}
+
+func (schema *OpenAPISchema) hasOneOf() bool {
+	return len(schema.inner.OneOf) > 0
+}
+
+func (schema *OpenAPISchema) oneOf() []Schema {
+	return schema.transformOpenAPISlice(schema.inner.OneOf)
+}
+
+func (schema *OpenAPISchema) requiredProperties() []string {
+	return schema.inner.Required
+}
+
+func (schema *OpenAPISchema) properties() map[string]Schema {
+	result := make(map[string]Schema)
+	for propName, propSchema := range schema.inner.Properties {
+		result[propName] = schema.withNewSchema(propSchema)
+	}
+
+	return result
+}
+
+func (schema *OpenAPISchema) description() *string {
+	if len(schema.inner.Description) == 0 {
+		return nil
+	}
+
+	return &schema.inner.Description
+}
+
+func (schema *OpenAPISchema) items() []Schema {
+	if schema.inner.Items.Schema != nil {
+		return []Schema{schema.withNewSchema(*schema.inner.Items.Schema)}
+	}
+
+	return schema.transformOpenAPISlice(schema.inner.Items.Schemas)
+}
+
+func (schema *OpenAPISchema) additionalPropertiesAllowed() bool {
+	return schema.inner.AdditionalProperties == nil || schema.inner.AdditionalProperties.Allows
+}
+
+func (schema *OpenAPISchema) additionalPropertiesSchema() Schema {
+	if schema.inner.AdditionalProperties == nil {
+		return nil
+	}
+
+	result := schema.inner.AdditionalProperties.Schema
+	if result == nil {
+		return nil
+	}
+
+	return schema.withNewSchema(*result)
+}
+
+func (schema *OpenAPISchema) enumValues() []string {
+	result := make([]string, len(schema.inner.Enum))
+	for i, enumValue := range schema.inner.Enum {
+		if enumString, ok := enumValue.(string); ok {
+			result[i] = fmt.Sprintf("%q", enumString)
+		} else if enumStringer, ok := enumValue.(fmt.Stringer); ok {
+			result[i] = fmt.Sprintf("%q", enumStringer.String())
+		} else if enumFloat, ok := enumValue.(float64); ok {
+			result[i] = fmt.Sprintf("%g", enumFloat)
+		} else {
+			panic(fmt.Sprintf("unable to convert enum value (%v %T) to string", enumValue, enumValue))
+		}
+	}
+
+	return result
+}
+
+func (schema *OpenAPISchema) isRef() bool {
+	return schema.inner.Ref.GetURL() != nil
+}
+
+// fetchFileRelative fetches the schema for the relative path created by combining 'baseFileName' and 'url'
```

Comment:
It's sorta confusing that you interleave the schema and cache stuff in the same file. It would help I think if they were in two different files, or were written one after the other (so `MakeCache` + `Cache` + `Cache` methods and then the schema methods)

Reply:
Really I want a module of the OpenApiSchema stuff (exporting just the schema/cache factories) and a module of the JSONSchema stuff (exporting just the schema factory) with only 1/2 files in each module, but Go feels like it doesn't encourage such small modules.

Reply:
But agree they were not meant to be interleaved, reordered to group all cache bits together.